[Writing Tools] iOS: Selection should be hidden when pondering effect begins, stay hidden until all animations complete

rr-codes opened the pull request:
#### 8c521f3355c750400b7344a99e0e6221867ad0e2
<pre>
[Writing Tools] iOS: Selection should be hidden when pondering effect begins, stay hidden until all animations complete
<a href="https://bugs.webkit.org/show_bug.cgi?id=278144">https://bugs.webkit.org/show_bug.cgi?id=278144</a>
<a href="https://rdar.apple.com/133427551">rdar://133427551</a>

Reviewed by Aditya Keerthi.

Apply the same logic as was previously done on macOS to ensure the selection is cleared during the animation.

Also, refactor the previously-macOS specific logic from WebViewImpl into WKWebView.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView compositionSession:didReceiveText:replacementRange:inContext:finished:]):
(-[WKWebView writingToolsSession:didReceiveAction:]):
(-[WKWebView _didEndPartialIntelligenceTextPonderingAnimation]):
(-[WKWebView _intelligenceTextPonderingAnimationIsComplete]):
(-[WKWebView _addTextAnimationForAnimationID:withData:]):
(-[WKWebView _removeTextAnimationForAnimationID:]):
(-[WKWebView _enableSourceTextAnimationAfterElementWithID:]):
(-[WKWebView _enableFinalTextAnimationForElementWithID:]):
(-[WKWebView _disableTextAnimationWithUUID:]):
(-[WKWebView _enableTextIndicatorStylingAfterElementWithID:]):
(-[WKWebView _enableTextIndicatorStylingForElementWithID:]):
(-[WKWebView _disableTextIndicatorStylingWithUUID:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didEndPartialIntelligenceTextPonderingAnimation):
(WebKit::PageClientImplCocoa::intelligenceTextPonderingAnimationIsComplete):
(WebKit::PageClientImplCocoa::addTextAnimationForAnimationID):
(WebKit::PageClientImplCocoa::removeTextAnimationForAnimationID):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didEndPartialIntelligenceTextPonderingAnimationImpl):
(WebKit::WebPageProxy::didEndPartialIntelligenceTextPonderingAnimation):
(WebKit::WebPageProxy::showSelectionForActiveWritingToolsSessionIfNeeded):
(WebKit::WebPageProxy::showSelectionForActiveWritingToolsSession): Deleted.
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView replacementEffectDidComplete]):
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager addTextAnimationForAnimationID:withData:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::writingToolsCompositionSessionDidReceiveRestartAction): Deleted.
(WebKit::WebViewImpl::writingToolsCompositionSessionDidReceiveReplacements): Deleted.
(WebKit::WebViewImpl::isWritingToolsTextReplacementsFinished const): Deleted.
(WebKit::WebViewImpl::isIntelligenceTextPonderingAnimationFinished const): Deleted.
(WebKit::WebViewImpl::willBeginPartialIntelligenceTextPonderingAnimation): Deleted.
(WebKit::WebViewImpl::didEndPartialIntelligenceTextPonderingAnimation): Deleted.
* Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift:
(TextAnimationManager.performReplacementAndGeneratePreview(for:effect:animation:)):
(TextAnimationManager.replacementEffectDidComplete(_:)):
* Source/WebKit/WebKitSwift/TextAnimation/WKSTextAnimationSourceDelegate.h:

Canonical link: <a href="https://commits.webkit.org/282570@main">https://commits.webkit.org/282570@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4216a58b1410d91460e1818361a19b8f456fdd46

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63464 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42820 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16061 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67485 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14072 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50508 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14352 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51103 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9727 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66533 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39752 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54959 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31790 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36431 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12944 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57955 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12644 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69181 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7411 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12227 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58408 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7443 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55038 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58643 "2 api tests failed or timed out") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6184 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9606 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38641 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39720 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40832 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39463 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift:
```diff
@@ -101,32 +101,36 @@ extension TextAnimationManager: UITextEffectView.ReplacementTextEffect.Delegate
         self.effectView.removeEffect(effect.id)
 
-        guard let (animationID, _) = chunkToEffect.first(where: { (_, value) in value == effect.id }) else {
+        guard let (animationID, _) = self.chunkToEffect.first(where: { (_, value) in value == effect.id }) else {
```
Swift noob question, what's the difference between the new and old line?

rr-codes replied:
It’s the same as other languages; it means the variable is an instance member variable and not a local variable. This just makes it more consistent throughout the file

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1211,11 +1211,21 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
 
 void WebPageProxy::writingToolsSessionDidReceiveAction(const WebCore::WritingTools::Session& session, WebCore::WritingTools::Action action)
 {
+    if (action == WebCore::WritingTools::Action::Restart && session.compositionType != WebCore::WritingTools::Session::CompositionType::SmartReply) {
```
Why is smart reply special cased here? Wouldn't `writingToolsTextReplacementsFinished` be in an incorrect state.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1211,11 +1211,21 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
 void WebPageProxy::compositionSessionDidReceiveTextWithReplacementRange(const WebCore::WritingTools::Session& session, const WebCore::AttributedString& attributedText, const WebCore::CharacterRange& range, const WebCore::WritingTools::Context& context, bool finished)
 {
+    internals().writingToolsTextReplacementsFinished = finished;
+    internals().partialIntelligenceTextPonderingAnimationCount += 1;
```
This needs to be reset if the process is terminated.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1288,17 +1298,31 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+    if (!hasRunningProcess())
+        return;
+
+    if (!internals().partialIntelligenceTextPonderingAnimationCount && internals().writingToolsTextReplacementsFinished) {
```
After reading this code, I think the state / mutation of these properties might make more sense on `WKWebView`. With a `PageClient` method to expose the result of this condition.

rr-codes committed:
[Writing Tools] iOS: Selection should be hidden when pondering effect begins, stay hidden until all animations complete
https://bugs.webkit.org/show_bug.cgi?id=278144
rdar://133427551

Reviewed by Aditya Keerthi.

Apply the same logic as was previously done on macOS to ensure the selection is cleared during the animation.

Also, refactor the previously-macOS specific logic from WebViewImpl into WKWebView.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView compositionSession:didReceiveText:replacementRange:inContext:finished:]):
(-[WKWebView writingToolsSession:didReceiveAction:]):
(-[WKWebView _didEndPartialIntelligenceTextPonderingAnimation]):
(-[WKWebView _intelligenceTextPonderingAnimationIsComplete]):
(-[WKWebView _addTextAnimationForAnimationID:withData:]):
(-[WKWebView _removeTextAnimationForAnimationID:]):
(-[WKWebView _enableSourceTextAnimationAfterElementWithID:]):
(-[WKWebView _enableFinalTextAnimationForElementWithID:]):
(-[WKWebView _disableTextAnimationWithUUID:]):
(-[WKWebView _enableTextIndicatorStylingAfterElementWithID:]):
(-[WKWebView _enableTextIndicatorStylingForElementWithID:]):
(-[WKWebView _disableTextIndicatorStylingWithUUID:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didEndPartialIntelligenceTextPonderingAnimation):
(WebKit::PageClientImplCocoa::intelligenceTextPonderingAnimationIsComplete):
(WebKit::PageClientImplCocoa::addTextAnimationForAnimationID):
(WebKit::PageClientImplCocoa::removeTextAnimationForAnimationID):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didEndPartialIntelligenceTextPonderingAnimationImpl):
(WebKit::WebPageProxy::didEndPartialIntelligenceTextPonderingAnimation):
(WebKit::WebPageProxy::showSelectionForActiveWritingToolsSessionIfNeeded):
(WebKit::WebPageProxy::showSelectionForActiveWritingToolsSession): Deleted.
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView replacementEffectDidComplete]):
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager addTextAnimationForAnimationID:withData:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::writingToolsCompositionSessionDidReceiveRestartAction): Deleted.
(WebKit::WebViewImpl::writingToolsCompositionSessionDidReceiveReplacements): Deleted.
(WebKit::WebViewImpl::isWritingToolsTextReplacementsFinished const): Deleted.
(WebKit::WebViewImpl::isIntelligenceTextPonderingAnimationFinished const): Deleted.
(WebKit::WebViewImpl::willBeginPartialIntelligenceTextPonderingAnimation): Deleted.
(WebKit::WebViewImpl::didEndPartialIntelligenceTextPonderingAnimation): Deleted.
* Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift:
(TextAnimationManager.performReplacementAndGeneratePreview(for:effect:animation:)):
(TextAnimationManager.replacementEffectDidComplete(_:)):
* Source/WebKit/WebKitSwift/TextAnimation/WKSTextAnimationSourceDelegate.h:

Canonical link: https://commits.webkit.org/282570@main
